src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs

```diff
@@ -96,157 +120,257 @@ public void Stop()
-                                    return;
+                                    this.logger.LogDebug("Process block from blocksQueue - '{0}'", this.chain.GetBlock(block.GetHash()));
+                                    await this.ProcessAsync(block).ConfigureAwait(false);
                                 }
 
-                                // Really ugly hack to let store catch up.
-                                // This will block the entire consensus pulling.
-                                this.logger.LogWarning("Wallet is behind the best chain and the next block is not found in store.");
-                                Thread.Sleep(100);
                                 continue;
                             }
+                            else
+                            {
+                                this.logger.LogDebug("Process block from blockStore - '{0}'", this.chain.GetBlock(nextBlock.GetHash()));
+                                await this.ProcessAsync(nextBlock).ConfigureAwait(false);
+                            }
 
                             break;
                         }
-
-                        this.walletTip = next;
-                        this.walletManager.ProcessBlock(nextblock, next);
                     }
                 }
                 else
                 {
-                    ChainedHeader findTip = this.walletTip.FindAncestorOrSelf(newTip);
-                    if (findTip == null)
+                    this.logger.LogDebug("(-)[TIP_HEIGHT_<_WALLET_TIP]");
+
+                    // Check if any Blocks are queued up.
+                    while (this.blocksQueue.TryDequeue(out Block block))
                     {
-                        this.logger.LogTrace("(-)[NEW_TIP_BEHIND_NOT_IN_WALLET]");
-                        return;
+                        this.logger.LogTrace("Process block from blocksQueue - '{0}'", this.chain.GetBlock(block.GetHash()));
+                        await this.ProcessAsync(block).ConfigureAwait(false);
                     }
-
-                    this.logger.LogTrace("Wallet tip '{0}' is ahead or equal to the new tip '{1}'.", this.walletTip, newTip);
                 }
             }
-            else this.logger.LogTrace("New block follows the previously known block '{0}'.", this.walletTip);
-
-            this.walletTip = newTip;
-            this.walletManager.ProcessBlock(block, newTip);
+            catch (OperationCanceledException)
+            {
+                this.logger.LogInformation("Stopping WalletSynceManager...");
+            }
+            catch (Exception e)
+            {
+                this.logger.LogError("Exception occurred: {0}", e.ToString());
+            }
 
             this.logger.LogTrace("(-)");
         }
 
         /// <inheritdoc />
-        public virtual void ProcessTransaction(Transaction transaction)
+        private void ProducerBlock(ITargetBlock<Block> target, Block block)
```

Comment:
`ProducerBlock` and `ConsumerBlockAsync` are 2 very small methods that are not used anywhere else but in `QueueBlock`. What's the advantage of splitting those in 3? I think the `QueueBlock` could include the other 2 and still be very readable.
In fact, `public void ProcessBlock(Block block)` only calls `QueueBlock` so it could be inlined there as well. 4 methods could be 1 without any loss of readability and a lot less logging :-)

Also the names `ProducerBlock` and `ConsumerBlockAsync` are not excessively descriptive.